File: Python/0110-e-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isBalanced(self, root: TreeNode) -> bool:
        if not root:
            return True
        
        left = self.height(root.left)
        right = self.height(root.right)
        
        if abs(left - right) <= 1:
            return self.isBalanced(root.left) and self.isBalanced(root.right)
        else: 
            return False
    
    def height(self, root):
        if not root:
            return 0
        
        left = self.height(root.left)
        right = self.height(root.right)
        
        return max(left, right) + 1 

File: leetcode/Java/0110-e-balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isBalanced(TreeNode root) {
        if (root == null) 
            return true;
        
        int left = getHeight(root.left);
        int right = getHeight(root.right);
        
        if (Math.abs(left - right) <= 1)
            return isBalanced(root.left) && isBalanced(root.right);
        else
            return false;
        
    }
    
    public int getHeight(TreeNode node) {
        if (node == null)
            return 0;
        
        int left = getHeight(node.left);
        int right = getHeight(node.right);
        
        return Math.max(left, right) + 1;
    }
}